### PR TITLE
Fix scalar broadcast

### PR DESCRIPTION
**Pull request by niboshi**
Fix: #779

I refactored `test_arithmetic` to test against Python scalars as well.

### CODE REVIEW

File: tests/cupy_tests/math_tests/test_arithmetic.py [hunk 1]
```diff
@@ -66,66 +46,163 @@ def check_raises_with_numpy_input(self, nargs, name):
...
+                    and isinstance(arg2, complex)):
+                y = y.astype(numpy.complex64)
 
-    def test_floor_divide(self):
-        with testing.NumpyError(divide='ignore'):
-            self.check_binary('floor_divide', no_complex=True)
-        self.check_raises_with_numpy_input(2, 'floor_divide')
+        # TODO(niboshi): Fix this. Treat NaN and inf interchangeably
+        y = xp.asarray(y)
+        if y.dtype in (float_types + complex_types):
```
Comment (unnonouno) [inline tests/cupy_tests/math_tests/test_arithmetic.py]: What does it mean?

Comment (niboshi) [reply tests/cupy_tests/math_tests/test_arithmetic.py]: Sorry, it was not clear. I added comments, and also narrowed the condition.

Comment (unnonouno) [reply tests/cupy_tests/math_tests/test_arithmetic.py]: Ok. Actually this behavior depends on CPU architecture I think (I mean ARM may returns different result for example).

**Comment by niboshi** [comment]
Fixed the error for NumPy<1.11.

File: tests/cupy_tests/math_tests/test_arithmetic.py (hunk 1 above)
Comment (niboshi) [reply tests/cupy_tests/math_tests/test_arithmetic.py]: I see. So you mean we don't have to fix this and we can just ignore the difference in the test?

Comment (unnonouno) [reply tests/cupy_tests/math_tests/test_arithmetic.py]: IMO, no because in ARM architecture for example TX1, this test fails maybe.

Comment (niboshi) [reply tests/cupy_tests/math_tests/test_arithmetic.py]: I'm sorry, it was not clear. I meant there are 2 choices and I think we should choose 2. Is that correct?

1. Fix `cupy.divide` to output `nan` or `inf` according to the CPU architecture
2. Do not fix `cupy.divide`. Instead fix the test to ignore the difference.

Comment (unnonouno) [reply tests/cupy_tests/math_tests/test_arithmetic.py]: Yes 2 is better i think. IMO, these behavior is undefined. So it is ok to skip its test.

**Comment by niboshi** [comment]
Fixed the comment to remove TODO.
Also I moved `xp.asarray` inside the condition.

**Comment by unnonouno** [comment]
jenkins test this pleasse

**Comment by niboshi** [comment]
jenkins test this pleasse

**Comment by niboshi** [comment]
jenkins test this pleasse

**Comment by kmaehashi** [comment]
Jenkins test this please

File: cupy/core/elementwise.pxi [hunk 2]
```diff
@@ -99,7 +99,10 @@ cdef dict _python_type_to_numpy_type = {
 
 
 cpdef _python_scalar_to_numpy_scalar(x):
-    if isinstance(x, six.integer_types):
+    # Note that isinstance(x, six_integer_types) matches with bool.
+    if isinstance(x, bool):
+        numpy_type = numpy.bool_
+    elif isinstance(x, six.integer_types):
```
Comment (unnonouno) [inline cupy/core/elementwise.pxi]: I think `type(x) in six.integer_types` is enough.

**Comment by niboshi** [comment]
I see. I'll fix as such.

By the way, the test is failing because NumPy behaves differently depending on the version.
How can I handle such version difference?
NumPy doesn't seem to provide tuple version number.

File: tests/cupy_tests/math_tests/test_arithmetic.py [hunk 3]
```diff
@@ -66,66 +46,169 @@ def check_raises_with_numpy_input(self, nargs, name):
...
-            self.check_binary('true_divide')
-        self.check_raises_with_numpy_input(2, 'true_divide')
+            elif (isinstance(arg1, (float, complex))
+                    and self.name in ('real', 'imag')):
+                y = xp.asarray(y)
 
-    def test_true_divide_negative(self):
-        with testing.NumpyError(divide='ignore'):
-            self.check_binary_negative('true_divide')
+        # TODO(niboshi): Fix this
```
Comment (unnonouno) [inline tests/cupy_tests/math_tests/test_arithmetic.py]: In numpy<1.11 `numpy.real(True)` returns bool.
```
>>> numpy.real(True)
array(True, dtype=bool)
```

**Comment by unnonouno** [comment]
The older numpy behaves differently. CuPy must behaves like the latest version of NumPy. I think we can skip the test in such cases.

**Comment by niboshi** [comment]
I agree. But how can I implement it?

If we convert NumPy result to int if it was bool, we cannot be sure if it's aligned to the latest NumPy version. (ex. In the future NumPy can revert its behavior to return bool.)

If we skip these binary tests entirely, I think we can't say we're supporting older versions.

**Comment by unnonouno** [comment]
How about checking numpy version and return types like
```
if numpy.__version__ < 1.11 and isinstance(return_value, bool):
    skip
```

**Comment by niboshi** [comment]
Yes, that's exactly what I want do. But isn't that code invalid?

**Comment by okuta** [comment]
jenkins, test this please.

File: cupy/core/elementwise.pxi (hunk 2 above)
Comment (niboshi) [reply cupy/core/elementwise.pxi]: Hmm, Is it worthwhile?
In my environment, both methods take almost exactly the same time (~93ns), and `isinstance` is more versatile and pythonic.

**Comment by unnonouno** [comment]
LGTM!!

**Pull request merged**